operator/pkg/helmreconciler/common.go

```diff
@@ -125,5 +126,77 @@ func applyOverlay(current, overlay runtime.Object) error {
 	if err != nil {
 		return err
 	}
-	return runtime.DecodeInto(unstructured.UnstructuredJSONScheme, merged, current)
+	// save any immutable values
+	clusterIP := getPath(current, name.ServiceStr, util.PathFromString("spec.clusterIP"))
+
+	err = runtime.DecodeInto(unstructured.UnstructuredJSONScheme, merged, current)
+	uc := current.UnstructuredContent()
+	uo := overlay.UnstructuredContent()
+	writeMap(uc, uo, util.PathFromString("spec"))
+	writeMap(uc, uo, util.PathFromString("metadata.labels"))
+
+	// restore any immutable values
+	writePath(current, name.ServiceStr, util.PathFromString("spec.clusterIP"), clusterIP)
+
+	return err
+
+}
+
+// writeMap replaces base with overlay at the given path. If the paths hit any nil non-leaf node, nothing is modified.
+func writeMap(base, overlay map[string]interface{}, path util.Path) {
+	for ; len(path) > 1; path = path[1:] {
+		pe := path[0]
+		if base[pe] == nil || overlay[pe] == nil {
+			return
+		}
+		bn, ok := base[pe].(map[string]interface{})
+		if !ok {
+			scope.Error("Unexpected type %T at path %s", base[pe], path)
```

Comment:
Is this expected to return something or `scope.Error` will be enough?

Reply:
this is not a normal error case, it would only pop up during development time if someone added an invalid k8s path.